File: services/connector-management-service/src/domain/egress.rs
use reqwest::Url;
use serde_json::Value;

use crate::AppState;

#[derive(Debug, Clone)]
pub struct EgressPolicy {
    pub allowed_hosts: Vec<String>,
    pub blocked_hosts: Vec<String>,
    pub allow_private_networks: bool,
    pub allow_insecure_http: bool,
}

impl EgressPolicy {
    pub fn from_state_and_config(state: &AppState, config: &Value) -> Self {
        let egress = config.get("egress");
        Self {
            allowed_hosts: string_list(egress.and_then(|value| value.get("allowed_hosts")))
                .or_else(|| {
                    (!state.allowed_egress_hosts.is_empty())
                        .then(|| state.allowed_egress_hosts.clone())
                })
                .unwrap_or_default(),
            blocked_hosts: string_list(egress.and_then(|value| value.get("blocked_hosts")))
                .unwrap_or_default(),
            allow_private_networks: egress
                .and_then(|value| value.get("allow_private_networks"))
                .and_then(Value::as_bool)
                .unwrap_or(state.allow_private_network_egress),
            allow_insecure_http: egress
                .and_then(|value| value.get("allow_insecure_http"))
                .and_then(Value::as_bool)
                .unwrap_or(false),
        }
    }
}

pub fn validate_url(url: &Url, policy: &EgressPolicy) -> Result<(), String> {
    let host = url
        .host_str()
        .ok_or_else(|| format!("URL '{}' has no host", url))?
        .to_ascii_lowercase();

    if !policy.allow_insecure_http
        && url.scheme() == "http"
        && !is_localhost(&host)
        && !is_private_ip(&host)
    {
        return Err(format!(
            "egress policy blocks insecure HTTP to host '{host}'"
        ));
    }

    if !policy.allow_private_networks && is_private_ip(&host) && !is_localhost(&host) {
        return Err(format!(
            "egress policy blocks private-network host '{host}'"
        ));
    }

    if host_matches_any(&host, &policy.blocked_hosts) {
        return Err(format!("egress policy blocks host '{host}'"));
    }

    if !policy.allowed_hosts.is_empty() && !host_matches_any(&host, &policy.allowed_hosts) {
        return Err(format!("egress policy does not allow host '{host}'"));
    }

    Ok(())
}

fn string_list(value: Option<&Value>) -> Option<Vec<String>> {
    value.and_then(Value::as_array).map(|items| {
        items
            .iter()
            .filter_map(Value::as_str)
            .map(|item| item.trim().to_ascii_lowercase())
            .filter(|item| !item.is_empty())
            .collect::<Vec<_>>()
    })
}

fn host_matches_any(host: &str, patterns: &[String]) -> bool {
    patterns.iter().any(|pattern| host_matches(host, pattern))
}

fn host_matches(host: &str, pattern: &str) -> bool {
    let normalized = pattern.trim().to_ascii_lowercase();
    if normalized.is_empty() {
        return false;
    }
    if normalized == "*" {
        return true;
    }
    if let Some(suffix) = normalized.strip_prefix("*.") {
        return host == suffix || host.ends_with(&format!(".{suffix}"));
    }
    host == normalized || host.ends_with(&format!(".{normalized}"))
}

fn is_localhost(host: &str) -> bool {
    matches!(host, "localhost" | "127.0.0.1" | "::1")
}

fn is_private_ip(host: &str) -> bool {
    let Ok(ip) = host.parse::<std::net::IpAddr>() else {
        return false;
    };
    match ip {
        std::net::IpAddr::V4(v4) => {
            let octets = v4.octets();
            octets[0] == 10
                || (octets[0] == 172 && (16..=31).contains(&octets[1]))
                || (octets[0] == 192 && octets[1] == 168)
                || octets[0] == 127
                || (octets[0] == 169 && octets[1] == 254)
        }
        std::net::IpAddr::V6(v6) => v6.is_loopback() || v6.is_unique_local(),
    }
}

#[cfg(test)]
mod tests {
    use reqwest::Url;

    use super::{EgressPolicy, validate_url};

    #[test]
    fn blocks_hosts_outside_allowlist() {
        let policy = EgressPolicy {
            allowed_hosts: vec!["api.example.com".to_string()],
            blocked_hosts: Vec::new(),
            allow_private_networks: false,
            allow_insecure_http: false,
        };
        let url = Url::parse("https://internal.example.net/v1/data").expect("url");
        assert!(validate_url(&url, &policy).is_err());
    }

    #[test]
    fn allows_wildcard_subdomains() {
        let policy = EgressPolicy {
            allowed_hosts: vec!["*.example.com".to_string()],
            blocked_hosts: Vec::new(),
            allow_private_networks: false,
            allow_insecure_http: false,
        };
        let url = Url::parse("https://sales.eu.example.com/odata").expect("url");
        assert!(validate_url(&url, &policy).is_ok());
    }

    #[test]
    fn blocks_private_network_when_disabled() {
        let policy = EgressPolicy {
            allowed_hosts: Vec::new(),
            blocked_hosts: Vec::new(),
            allow_private_networks: false,
            allow_insecure_http: true,
        };
        let url = Url::parse("http://10.0.0.4/feed").expect("url");
        assert!(validate_url(&url, &policy).is_err());
    }
}

File: services/connector-management-service/internal/domain/egress/egress.go
// Package egress mirrors the Rust connector-management egress policy helpers in
// `services/connector-management-service/src/domain/egress.rs`.
//
// The helpers only validate URLs against policy metadata owned by the connector
// management plane. They do not open sockets, resolve DNS, or enforce an
// external network boundary; Rust delegates that boundary to the HTTP runtime /
// connector-agent / network-boundary services, and Go keeps the same ownership
// split.
package egress

import (
	"fmt"
	"net"
	"net/url"
	"strings"
)

// Policy mirrors Rust's EgressPolicy. Host lists are normalized to trimmed,
// lower-case values by FromStateAndConfig.
type Policy struct {
	AllowedHosts         []string
	BlockedHosts         []string
	AllowPrivateNetworks bool
	AllowInsecureHTTP    bool
}

// FromStateAndConfig mirrors Rust's EgressPolicy::from_state_and_config.
//
// The state inputs correspond to AppState.allowed_egress_hosts and
// AppState.allow_private_network_egress. Per-connection overrides live under
// config.egress.{allowed_hosts, blocked_hosts, allow_private_networks,
// allow_insecure_http}. Non-string entries in host lists are ignored, matching
// Rust's serde_json::Value::as_str filter_map behavior.
func FromStateAndConfig(allowedHosts []string, allowPrivateNetworks bool, config map[string]any) Policy {
	policy := Policy{
		AllowedHosts:         normalizeHostList(allowedHosts),
		BlockedHosts:         nil,
		AllowPrivateNetworks: allowPrivateNetworks,
		AllowInsecureHTTP:    false,
	}
	egressConfig, _ := config["egress"].(map[string]any)
	if egressConfig == nil {
		return policy
	}
	if hosts := stringList(egressConfig["allowed_hosts"]); hosts != nil {
		policy.AllowedHosts = hosts
	}
	if hosts := stringList(egressConfig["blocked_hosts"]); hosts != nil {
		policy.BlockedHosts = hosts
	}
	if allow, ok := egressConfig["allow_private_networks"].(bool); ok {
		policy.AllowPrivateNetworks = allow
	}
	if allow, ok := egressConfig["allow_insecure_http"].(bool); ok {
		policy.AllowInsecureHTTP = allow
	}
	return policy
}

// ValidateURL mirrors Rust's validate_url.
func ValidateURL(u *url.URL, policy Policy) error {
	if u == nil {
		return fmt.Errorf("URL '<nil>' has no host")
	}
	host := strings.ToLower(u.Hostname())
	if host == "" {
		return fmt.Errorf("URL '%s' has no host", u.String())
	}

	scheme := strings.ToLower(u.Scheme)
	if !policy.AllowInsecureHTTP && scheme == "http" && !IsLocalhost(host) && !IsPrivateIP(host) {
		return fmt.Errorf("egress policy blocks insecure HTTP to host '%s'", host)
	}
	if !policy.AllowPrivateNetworks && IsPrivateIP(host) && !IsLocalhost(host) {
		return fmt.Errorf("egress policy blocks private-network host '%s'", host)
	}
	if HostMatchesAny(host, policy.BlockedHosts) {
		return fmt.Errorf("egress policy blocks host '%s'", host)
	}
	if len(policy.AllowedHosts) > 0 && !HostMatchesAny(host, policy.AllowedHosts) {
		return fmt.Errorf("egress policy does not allow host '%s'", host)
	}
	return nil
}

func stringList(value any) []string {
	items, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(items))
	for _, item := range items {
		s, ok := item.(string)
		if !ok {
			continue
		}
		s = strings.ToLower(strings.TrimSpace(s))
		if s != "" {
			out = append(out, s)
		}
	}
	return out
}

func normalizeHostList(hosts []string) []string {
	if len(hosts) == 0 {
		return nil
	}
	out := make([]string, 0, len(hosts))
	for _, host := range hosts {
		host = strings.ToLower(strings.TrimSpace(host))
		if host != "" {
			out = append(out, host)
		}
	}
	return out
}

// HostMatchesAny returns true when host matches any allow/block pattern.
func HostMatchesAny(host string, patterns []string) bool {
	for _, pattern := range patterns {
		if HostMatches(host, pattern) {
			return true
		}
	}
	return false
}

// HostMatches mirrors Rust's host_matches helper.
func HostMatches(host, pattern string) bool {
	host = strings.ToLower(host)
	pattern = strings.ToLower(strings.TrimSpace(pattern))
	if pattern == "" {
		return false
	}
	if pattern == "*" {
		return true
	}
	if suffix, ok := strings.CutPrefix(pattern, "*."); ok {
		return host == suffix || strings.HasSuffix(host, "."+suffix)
	}
	return host == pattern || strings.HasSuffix(host, "."+pattern)
}

// IsLocalhost mirrors Rust's is_localhost helper.
func IsLocalhost(host string) bool {
	switch strings.ToLower(host) {
	case "localhost", "127.0.0.1", "::1":
		return true
	default:
		return false
	}
}

// IsPrivateIP mirrors Rust's is_private_ip helper: 10/8, 172.16/12,
// 192.168/16, 127/8, 169.254/16, IPv6 loopback, and IPv6 unique-local
// addresses (fc00::/7) are private. Hostnames are not resolved here.
func IsPrivateIP(host string) bool {
	ip := net.ParseIP(host)
	if ip == nil {
		return false
	}
	if v4 := ip.To4(); v4 != nil {
		switch {
		case v4[0] == 10:
			return true
		case v4[0] == 172 && v4[1] >= 16 && v4[1] <= 31:
			return true
		case v4[0] == 192 && v4[1] == 168:
			return true
		case v4[0] == 127:
			return true
		case v4[0] == 169 && v4[1] == 254:
			return true
		default:
			return false
		}
	}
	if ip.IsLoopback() {
		return true
	}
	return len(ip) == net.IPv6len && (ip[0]&0xfe) == 0xfc
}
